File: sample/nim/03_julia_set.nim
include "py_runtime.nim"

import std/os, std/times, std/tables, std/strutils, std/math

proc render_julia*(width: int, height: int, max_iter: int, cx: float, cy: float): auto =
  var pixels: seq[uint8] = newSeq[uint8]()
  var v_hoisted_cast_1: float = float((height - 1))
  var v_hoisted_cast_2: float = float((width - 1))
  var v_hoisted_cast_3: float = float(max_iter)
  for y in 0 ..< height:
    var zy0: float = ((-1.2) + (2.4 * (float(y) / float(v_hoisted_cast_1))))
    for x in 0 ..< width:
      var zx: float = ((-1.8) + (3.6 * (float(x) / float(v_hoisted_cast_2))))
      var zy: float = zy0
      var i: int = 0
      while (i < max_iter):
        var zx2: float = (zx * zx)
        var zy2: float = (zy * zy)
        if ((zx2 + zy2) > 4.0):
          discard `break`
        zy = (((2.0 * zx) * zy) + cy)
        zx = ((zx2 - zy2) + cx)
        i += 1
      var r: int = 0
      var g: int = 0
      var b: int = 0
      if (i >= max_iter):
        r = 0
        g = 0
        b = 0
      else:
        var t: float = (float(i) / float(v_hoisted_cast_3))
        r = int((255.0 * (0.2 + (0.8 * t))))
        g = int((255.0 * (0.1 + (0.9 * (t * t)))))
        b = int((255.0 * (1.0 - t)))
      pixels.add(uint8(r))
      pixels.add(uint8(g))
      pixels.add(uint8(b))
  return pixels

proc run_julia*(): auto =
  var width: int = 3840
  var height: int = 2160
  var max_iter: int = 20000
  var out_path: string = "sample/out/03_julia_set.png"
  var start: float = epochTime()
  var pixels: seq[uint8] = render_julia(width, height, max_iter, (-0.8), 0.156)
  write_rgb_png(out_path, width, height, pixels)
  var elapsed: float = (epochTime() - start)
  echo "output:", out_path
  echo "size:", width, "x", height
  echo "max_iter:", max_iter
  echo "elapsed_sec:", elapsed


if isMainModule:
  run_julia()

File: sample/dart/03_julia_set.dart
import 'py_runtime.dart';
import 'png/east.dart' as png;


var perf_counter = pytraPerfCounter;

// --- pytra runtime helpers ---
String __pytraPrintRepr(dynamic v) {
  if (v == true) return 'True';
  if (v == false) return 'False';
  if (v == null) return 'None';
  return v.toString();
}

void __pytraPrint(List<dynamic> args) {
  print(args.map(__pytraPrintRepr).join(' '));
}

bool __pytraTruthy(dynamic v) {
  if (v == null) return false;
  if (v is bool) return v;
  if (v is num) return v != 0;
  if (v is String) return v.isNotEmpty;
  if (v is List) return v.isNotEmpty;
  if (v is Map) return v.isNotEmpty;
  return true;
}

bool __pytraContains(dynamic container, dynamic value) {
  if (container is List) return container.contains(value);
  if (container is Map) return container.containsKey(value);
  if (container is Set) return container.contains(value);
  if (container is String) return container.contains(value.toString());
  return false;
}

dynamic __pytraRepeatSeq(dynamic a, dynamic b) {
  dynamic seq = a;
  dynamic count = b;
  if (a is num && b is! num) { seq = b; count = a; }
  int n = (count is num) ? count.toInt() : 0;
  if (n <= 0) {
    if (seq is String) return '';
    return [];
  }
  if (seq is String) return seq * n;
  if (seq is List) {
    var out = [];
    for (var i = 0; i < n; i++) { out.addAll(seq); }
    return out;
  }
  return (a is num ? a : 0) * (b is num ? b : 0);
}

bool __pytraStrIsdigit(String s) {
  if (s.isEmpty) return false;
  for (var i = 0; i < s.length; i++) {
    var c = s.codeUnitAt(i);
    if (c < 48 || c > 57) return false;
  }
  return true;
}

bool __pytraStrIsalpha(String s) {
  if (s.isEmpty) return false;
  for (var i = 0; i < s.length; i++) {
    var c = s.codeUnitAt(i);
    if (!((c >= 65 && c <= 90) || (c >= 97 && c <= 122))) return false;
  }
  return true;
}

bool __pytraStrIsalnum(String s) {
  if (s.isEmpty) return false;
  for (var i = 0; i < s.length; i++) {
    var c = s.codeUnitAt(i);
    if (!((c >= 48 && c <= 57) || (c >= 65 && c <= 90) || (c >= 97 && c <= 122))) return false;
  }
  return true;
}
// --- end runtime helpers ---

// 03: Sample that outputs a Julia set as a PNG image.
// Implemented with simple loop-centric logic for transpilation compatibility.

dynamic render_julia(int width, int height, int max_iter, double cx, double cy) {
  var pixels = <int>[];
  
  for (var y = 0; y < height; y++) {
    double zy0 = ((-1.2) + (2.4 * (y / (height - 1))));
    
    for (var x = 0; x < width; x++) {
      double zx = ((-1.8) + (3.6 * (x / (width - 1))));
      double zy = zy0;
      
      int i = 0;
      while ((i < max_iter)) {
        double zx2 = (zx * zx);
        double zy2 = (zy * zy);
        if (((zx2 + zy2) > 4.0)) {
          break;
        }
        zy = (((2.0 * zx) * zy) + cy);
        zx = ((zx2 - zy2) + cx);
        i = ((i + 1) as int);
      }
      int r = 0;
      int g = 0;
      int b = 0;
      if ((i >= max_iter)) {
        r = 0;
        g = 0;
        b = 0;
      } else {
        double t = (i / max_iter);
        r = pytraInt((255.0 * (0.2 + (0.8 * t))));
        g = pytraInt((255.0 * (0.1 + (0.9 * (t * t)))));
        b = pytraInt((255.0 * (1.0 - t)));
      }
      pixels.add(r);
      pixels.add(g);
      pixels.add(b);
    }
  }
  return pixels;
}

void run_julia() {
  int width = 3840;
  int height = 2160;
  int max_iter = 20000;
  String out_path = "sample/out/03_julia_set.png";
  
  double start = perf_counter();
  var pixels = render_julia(width, height, max_iter, (-0.8), 0.156);
  png.write_rgb_png(out_path, width, height, pixels);
  double elapsed = (perf_counter() - start);
  
  __pytraPrint(["output:", out_path]);
  __pytraPrint(["size:", width, "x", height]);
  __pytraPrint(["max_iter:", max_iter]);
  __pytraPrint(["elapsed_sec:", elapsed]);
}


void main() {
  run_julia();
}
